# Deep Learning Challenge
>* Section 5.3.6

* Using the fashion MNIST dataset:
* 1. Preprocess your data so that you can feed it into ANN models.
* 2. Split your data into training and test sets.
* 3. Try different ANN models and train them on your training set. You can play with the following:

>* Number of layers
>* Activation functions of the layers
>* Number of neurons in the layers
>* Different batch sizes during training

* 4. Compare your models' training scores and interpret your results.
* 5. Evaluate how your models perform on your test set. Compare the results of your models. 

In [1]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense

In [2]:
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

In [3]:
print("Train images shape:", X_train.shape)
print("Train labels shape:", y_train.shape)
print("Test images shape:", X_test.shape)
print("Test labels shape:", y_test.shape)

Train images shape: (60000, 28, 28)
Train labels shape: (60000,)
Test images shape: (10000, 28, 28)
Test labels shape: (10000,)


In [4]:
input_dim = 784 
output_dim = nb_classes = 10
nb_epoch = 20

X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

In [5]:
Y_train = to_categorical(y_train, nb_classes)
Y_test = to_categorical(y_test, nb_classes)

In [6]:
model = Sequential()
# The first dense layer
model.add(Dense(128, input_shape=(784,), activation="tanh"))
# The second dense layer
model.add(Dense(64, activation="tanh"))
# Third dense layer
model.add(Dense(32, activation="tanh"))
# The last layer is the output layer
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

model.fit(X_train, Y_train, batch_size=8, epochs=20, verbose=1)

Epoch 1/20
7500/7500 [==============================] - 12s 2ms/step - loss: 0.5429 - accuracy: 0.8126
Epoch 2/20
7500/7500 [==============================] - 12s 2ms/step - loss: 0.4030 - accuracy: 0.8541
Epoch 3/20
7500/7500 [==============================] - 12s 2ms/step - loss: 0.3675 - accuracy: 0.8670
Epoch 4/20
7500/7500 [==============================] - 11s 2ms/step - loss: 0.3426 - accuracy: 0.8753
Epoch 5/20
7500/7500 [==============================] - 11s 2ms/step - loss: 0.3271 - accuracy: 0.8806
Epoch 6/20
7500/7500 [==============================] - 12s 2ms/step - loss: 0.3122 - accuracy: 0.8859
Epoch 7/20
7500/7500 [==============================] - 11s 2ms/step - loss: 0.2999 - accuracy: 0.8906
Epoch 8/20
7500/7500 [==============================] - 12s 2ms/step - loss: 0.2895 - accuracy: 0.8937
Epoch 9/20
7500/7500 [==============================] - 12s 2ms/step - loss: 0.2806 - accuracy: 0.8977
Epoch 10/20
7500/7500 [==============================] - 12s 2ms/step - l

In [7]:
train_score = model.evaluate(X_train, Y_train, verbose=0)
print('Training score:', train_score[0])
print('Training accuracy:', train_score[1])

Training score: 0.21143603324890137
Training accuracy: 0.9205333590507507


In [8]:
model = Sequential()
# The first dense layer
model.add(Dense(128, input_shape=(784,), activation="tanh"))
# The second dense layer
model.add(Dense(64, activation="tanh"))
# The last layer is the output layer
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

model.fit(X_train, Y_train, batch_size=8, epochs=20, verbose=1)

Epoch 1/20
7500/7500 [==============================] - 11s 2ms/step - loss: 0.5392 - accuracy: 0.8096
Epoch 2/20
7500/7500 [==============================] - 11s 1ms/step - loss: 0.4061 - accuracy: 0.8535
Epoch 3/20
7500/7500 [==============================] - 11s 1ms/step - loss: 0.3727 - accuracy: 0.8659
Epoch 4/20
7500/7500 [==============================] - 12s 2ms/step - loss: 0.3511 - accuracy: 0.8726
Epoch 5/20
7500/7500 [==============================] - 11s 1ms/step - loss: 0.3339 - accuracy: 0.8780
Epoch 6/20
7500/7500 [==============================] - 11s 1ms/step - loss: 0.3205 - accuracy: 0.8827
Epoch 7/20
7500/7500 [==============================] - 11s 1ms/step - loss: 0.3078 - accuracy: 0.8877
Epoch 8/20
7500/7500 [==============================] - 11s 1ms/step - loss: 0.2968 - accuracy: 0.8893
Epoch 9/20
7500/7500 [==============================] - 11s 1ms/step - loss: 0.2867 - accuracy: 0.8942
Epoch 10/20
7500/7500 [==============================] - 11s 1ms/step - l

In [9]:
train_score = model.evaluate(X_train, Y_train, verbose=0)
print('Training score:', train_score[0])
print('Training accuracy:', train_score[1])

Training score: 0.200212761759758
Training accuracy: 0.9269833564758301


In [10]:
model = Sequential()
# The first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# The second dense layer
model.add(Dense(64, activation="relu"))
# Third dense layer
model.add(Dense(32, activation="relu"))
# The last layer is the output layer
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

model.fit(X_train, Y_train, batch_size=8, epochs=20, verbose=1)

Epoch 1/20
7500/7500 [==============================] - 12s 2ms/step - loss: 0.5868 - accuracy: 0.7950
Epoch 2/20
7500/7500 [==============================] - 12s 2ms/step - loss: 0.4086 - accuracy: 0.8519
Epoch 3/20
7500/7500 [==============================] - 12s 2ms/step - loss: 0.3670 - accuracy: 0.8641
Epoch 4/20
7500/7500 [==============================] - 12s 2ms/step - loss: 0.3422 - accuracy: 0.8734
Epoch 5/20
7500/7500 [==============================] - 12s 2ms/step - loss: 0.3228 - accuracy: 0.8813
Epoch 6/20
7500/7500 [==============================] - 12s 2ms/step - loss: 0.3088 - accuracy: 0.8854
Epoch 7/20
7500/7500 [==============================] - 12s 2ms/step - loss: 0.2951 - accuracy: 0.8909
Epoch 8/20
7500/7500 [==============================] - 12s 2ms/step - loss: 0.2843 - accuracy: 0.8935
Epoch 9/20
7500/7500 [==============================] - 12s 2ms/step - loss: 0.2750 - accuracy: 0.8969
Epoch 10/20
7500/7500 [==============================] - 12s 2ms/step - l

In [11]:
train_score = model.evaluate(X_train, Y_train, verbose=0)
print('Training score:', train_score[0])
print('Training accuracy:', train_score[1])

Training score: 0.19804686307907104
Training accuracy: 0.92535001039505


In [12]:
model = Sequential()
# The first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# The second dense layer
model.add(Dense(64, activation="relu"))
# The last layer is the output layer
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

model.fit(X_train, Y_train, batch_size=8, epochs=20, verbose=1)

Epoch 1/20
7500/7500 [==============================] - 11s 1ms/step - loss: 0.5686 - accuracy: 0.8025
Epoch 2/20
7500/7500 [==============================] - 10s 1ms/step - loss: 0.4112 - accuracy: 0.8527
Epoch 3/20
7500/7500 [==============================] - 11s 1ms/step - loss: 0.3674 - accuracy: 0.8677
Epoch 4/20
7500/7500 [==============================] - 11s 1ms/step - loss: 0.3407 - accuracy: 0.8766
Epoch 5/20
7500/7500 [==============================] - 11s 1ms/step - loss: 0.3224 - accuracy: 0.8808
Epoch 6/20
7500/7500 [==============================] - 11s 1ms/step - loss: 0.3079 - accuracy: 0.8865
Epoch 7/20
7500/7500 [==============================] - 11s 1ms/step - loss: 0.2957 - accuracy: 0.8915
Epoch 8/20
7500/7500 [==============================] - 11s 1ms/step - loss: 0.2847 - accuracy: 0.8942
Epoch 9/20
7500/7500 [==============================] - 11s 1ms/step - loss: 0.2757 - accuracy: 0.8987
Epoch 10/20
7500/7500 [==============================] - 11s 1ms/step - l

In [13]:
train_score = model.evaluate(X_train, Y_train, verbose=0)
print('Training score:', train_score[0])
print('Training accuracy:', train_score[1])

Training score: 0.18942220509052277
Training accuracy: 0.9304500222206116


In [14]:
model = Sequential()
# The first dense layer
model.add(Dense(128, input_shape=(784,), activation="tanh"))
# The second dense layer
model.add(Dense(64, activation="tanh"))
# Third dense layer
model.add(Dense(32, activation="tanh"))
# The last layer is the output layer
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 1s 2ms/step - loss: 1.1205 - accuracy: 0.6589
Epoch 2/20
469/469 [==============================] - 1s 2ms/step - loss: 0.7038 - accuracy: 0.7770
Epoch 3/20
469/469 [==============================] - 1s 2ms/step - loss: 0.5898 - accuracy: 0.8076
Epoch 4/20
469/469 [==============================] - 1s 2ms/step - loss: 0.5323 - accuracy: 0.8211
Epoch 5/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4977 - accuracy: 0.8299
Epoch 6/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4744 - accuracy: 0.8365
Epoch 7/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4576 - accuracy: 0.8406
Epoch 8/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4436 - accuracy: 0.8450
Epoch 9/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4321 - accuracy: 0.8487
Epoch 10/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4223 - accuracy: 0.8518

In [15]:
train_score = model.evaluate(X_train, Y_train, verbose=0)
print('Training score:', train_score[0])
print('Training accuracy:', train_score[1])

Training score: 0.36468929052352905
Training accuracy: 0.870199978351593


In [16]:
model = Sequential()
# The first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# The second dense layer
model.add(Dense(64, activation="relu"))
# Third dense layer
model.add(Dense(32, activation="relu"))
# The last layer is the output layer
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 1s 2ms/step - loss: 1.2330 - accuracy: 0.5789
Epoch 2/20
469/469 [==============================] - 1s 2ms/step - loss: 0.6737 - accuracy: 0.7730
Epoch 3/20
469/469 [==============================] - 1s 2ms/step - loss: 0.5687 - accuracy: 0.8053
Epoch 4/20
469/469 [==============================] - 1s 2ms/step - loss: 0.5186 - accuracy: 0.8208
Epoch 5/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4912 - accuracy: 0.8291
Epoch 6/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4702 - accuracy: 0.8364
Epoch 7/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4559 - accuracy: 0.8405
Epoch 8/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4416 - accuracy: 0.8457
Epoch 9/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4309 - accuracy: 0.8499
Epoch 10/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4199 - accuracy: 0.8531

In [17]:
train_score = model.evaluate(X_train, Y_train, verbose=0)
print('Training score:', train_score[0])
print('Training accuracy:', train_score[1])

Training score: 0.3533790111541748
Training accuracy: 0.8759499788284302


In [18]:
model = Sequential()
# The first dense layer
model.add(Dense(156, input_shape=(784,), activation="tanh"))
# The second dense layer
model.add(Dense(78, activation="tanh"))
# Third dense layer
model.add(Dense(39, activation="tanh"))
# The last layer is the output layer
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 1s 2ms/step - loss: 1.0024 - accuracy: 0.7041
Epoch 2/20
469/469 [==============================] - 1s 2ms/step - loss: 0.6301 - accuracy: 0.7946
Epoch 3/20
469/469 [==============================] - 1s 2ms/step - loss: 0.5467 - accuracy: 0.8140
Epoch 4/20
469/469 [==============================] - 1s 2ms/step - loss: 0.5039 - accuracy: 0.8253
Epoch 5/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4774 - accuracy: 0.8326
Epoch 6/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4582 - accuracy: 0.8388
Epoch 7/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4434 - accuracy: 0.8436
Epoch 8/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4306 - accuracy: 0.8483
Epoch 9/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4205 - accuracy: 0.8511
Epoch 10/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4112 - accuracy: 0.8541

In [19]:
train_score = model.evaluate(X_train, Y_train, verbose=0)
print('Training score:', train_score[0])
print('Training accuracy:', train_score[1])

Training score: 0.3555021584033966
Training accuracy: 0.8713499903678894


In [20]:
model = Sequential()
# The first dense layer
model.add(Dense(156, input_shape=(784,), activation="relu"))
# The second dense layer
model.add(Dense(78, activation="relu"))
# Third dense layer
model.add(Dense(39, activation="relu"))
# The last layer is the output layer
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 1s 2ms/step - loss: 1.2530 - accuracy: 0.5946
Epoch 2/20
469/469 [==============================] - 1s 2ms/step - loss: 0.6499 - accuracy: 0.7823
Epoch 3/20
469/469 [==============================] - 1s 2ms/step - loss: 0.5525 - accuracy: 0.8118
Epoch 4/20
469/469 [==============================] - 1s 2ms/step - loss: 0.5068 - accuracy: 0.8249
Epoch 5/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4802 - accuracy: 0.8334
Epoch 6/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4613 - accuracy: 0.8399
Epoch 7/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4450 - accuracy: 0.8441
Epoch 8/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4348 - accuracy: 0.8481
Epoch 9/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4213 - accuracy: 0.8516
Epoch 10/20
469/469 [==============================] - 1s 2ms/step - loss: 0.4120 - accuracy: 0.8555

In [21]:
train_score = model.evaluate(X_train, Y_train, verbose=0)
print('Training score:', train_score[0])
print('Training accuracy:', train_score[1])

Training score: 0.34525319933891296
Training accuracy: 0.8782333135604858



```
| Activation Function | Layers          | Batch Size | Accuracy Score | Training Score |
|---------------------|-----------------|------------|----------------|----------------|
| Tahn                | 3 (128, 64, 32) | 8          | 0.92           | 0.21           |
| Tahn                | 2 (128, 64)     | 8          | 0.93           | 0.20           |
|                     |                 |            |                |                |
| Relu                | 3 (128, 64, 32) | 8          | 0.93           | 0.20           |
| Relu                | 2 (128, 64)     | 8          | 0.93           | 0.19           |
|                     |                 |            |                |                |
| Tahn                | 3 (128, 64, 32) | 128        | 0.87           | 0.36           |
| Relu                | 3 (128, 64, 32) | 128        | 0.88           | 0.35           |
|                     |                 |            |                |                |
| Tahn                | 3 (156, 78, 39) | 128        | 0.87           | 0.36           |
| Relu                | 3 (156, 78, 39) | 128        | 0.88           | 0.35           |

* The models performed very similarly (Tahn & Relu) with the changes in parameters.
* The best models were 2 layer (128, 64, 10) with batch size 8.
* Their was a very slight loss of performance when using 2 layers.
<br>
* Both models again were very similar when the batch size was increased to 128. They did see a marked drop in performance.
* A change to the number of nuerons didn't seem to have an effect at batch size 128.
<br>
* With this data set a lower batch size was the main factor in scoring.